openpathsampling/beta/hooks.py

```diff
@@ -346,3 +360,92 @@ def after_simulation(self, sim, hook_state):
             refresh=False,
             output_stream=self.output_stream
         )
+
+
+class GraciousKillHook(PathSimulatorHook):
+    """
+    'Graciously' kill a simulation when the maximum walltime is reached.
+
+    If this hook is attached to PathSimulator, it will continously estimate
+    the runtime per step. After each step it checks if the next step would
+    exceed the maximum walltime, if so it syncs and closes the storage, then
+    it calls a custom/user provided function (if any) and finally raises a
+    'GraciousKillError' to end the simulation loop.
+
+    Example usage
+    -------------
+    ```
+    kill_hook = GraciousKillHook("3 minutes 34 seconds")
+    # sampler is a `PathSimulator`
+    sampler.attach_hook(kill_hook)
+    try:
+        sampler.run(2000)
+    except GraciousKillError:
+        print("Simulation ended due to maximum walltime reached")
+    ```
+    """
+
+    implemented_for = ["before_simulation", "after_step"]
+
+    def __init__(self, max_walltime, fuzziness=1, final_call=None):
+        """
+        Initialize a GraciousKillHook.
+
+        Parameters
+        ----------
+        max_walltime - str, maximum allowed walltime,
+                       e.g. `23 hours 20 minutes`
+        fuzziness - float (default=0.9), fraction to add to the time per step
+                    when estimating if the next step could exceed the maximum
+                    walltime, i.e. the default is to stop when the time left
+                    suffices for slighlty less than 2 steps
+        final_call - None or callable (default None), will be called when the
+                     simulation is killed, it must take one argument, the hook
+                     passes the step number of the step after which it killed
+                     the simulation
+        """
+        self.max_walltime = max_walltime
+        self._timedelta = self._get_timedelta(max_walltime).total_seconds()
+        logger.info("Parsed time string '{:s}' as a ".format(self.max_walltime)
+                    + "timedelta of {:d} seconds.".format(int(self._timedelta))
+                    )
+        self.fuzziness = fuzziness
+        self.final_call = final_call
+
+    def _get_timedelta(self, time_str):
+        # https://stackoverflow.com/questions/35545091/convert-human-readable-time-difference-not-timestamp-to-something-usable-for-s/35545140
```

Comment:
```suggestion
        # https://stackoverflow.com/questions/35545091/
```

Equivalent link for the question, keeping to 80 chars. You might actually want to change the first part to `https://stackoverflow.com/a/35545140`, which is the answer that it looks like the following is based on.  (Was orig link supposed to be https://stackoverflow.com/questions/35545091/convert-human-readable-time-difference-not-timestamp-to-something-usable-for-s/#35545140 ?)